File: services/calc/debrief_calc/result_types.py
"""
Result type classification for debrief-calc tool responses.

Provides the four top-level result types and hierarchical type path matching.
"""

from enum import Enum


class ResultTopType(str, Enum):
    """Four permitted top-level result types."""
    MUTATION = "mutation"
    ADDITION = "addition"
    DELETION = "deletion"
    ARTIFACT = "artifact"


class ResultTypePath:
    """Hierarchical result type path (e.g., 'mutation/track/smoothed').

    The first segment must be a valid ResultTopType. Paths are slash-delimited
    with no empty segments, leading, or trailing slashes.
    """

    def __init__(self, path: str) -> None:
        if not path or not isinstance(path, str):
            raise ValueError("path must be a non-empty string")
        path = path.strip()
        if path.startswith("/") or path.endswith("/"):
            raise ValueError("path must not have leading or trailing slashes")
        self._segments = path.split("/")
        if any(s == "" for s in self._segments):
            raise ValueError("path must not contain empty segments")
        try:
            self._top_level = ResultTopType(self._segments[0])
        except ValueError as err:
            valid = ", ".join(t.value for t in ResultTopType)
            raise ValueError(
                f"first segment must be a valid ResultTopType ({valid}), got: '{self._segments[0]}'"
            ) from err
        self._path = path

    @property
    def path(self) -> str:
        return self._path

    @property
    def top_level(self) -> ResultTopType:
        return self._top_level

    @property
    def segments(self) -> list[str]:
        return list(self._segments)

    def matches(self, prefix: str) -> bool:
        """Check if this type path starts with the given prefix.

        Matching is segment-based: 'mutation' matches 'mutation/track/smoothed'
        but 'mut' does not.
        """
        if not prefix:
            return False
        prefix_segments = prefix.split("/")
        if len(prefix_segments) > len(self._segments):
            return False
        return self._segments[:len(prefix_segments)] == prefix_segments

    def __str__(self) -> str:
        return self._path

    def __repr__(self) -> str:
        return f"ResultTypePath('{self._path}')"

    def __eq__(self, other: object) -> bool:
        if isinstance(other, ResultTypePath):
            return self._path == other._path
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._path)

File: shared/components/diff/src/resultTypes.ts
/**
 * Result type matching utilities for hierarchical type degradation.
 *
 * Enables consumers to match result types at any depth in the hierarchy.
 */

export type ResultTopType = "mutation" | "addition" | "deletion" | "artifact";

const VALID_TOP_TYPES: ReadonlySet<string> = new Set<ResultTopType>([
  "mutation",
  "addition",
  "deletion",
  "artifact",
]);

/**
 * Match a result type path against a prefix.
 *
 * Matching is segment-based: "mutation" matches "mutation/track/smoothed"
 * but "mut" does not.
 */
export function matchesResultType(typePath: string, prefix: string): boolean {
  if (!typePath || !prefix) return false;
  const pathSegments = typePath.split("/");
  const prefixSegments = prefix.split("/");
  if (prefixSegments.length > pathSegments.length) return false;
  for (let i = 0; i < prefixSegments.length; i++) {
    if (pathSegments[i] !== prefixSegments[i]) return false;
  }
  return true;
}

/**
 * Extract the top-level type from a result type path.
 *
 * @throws Error if the type path doesn't start with a valid top-level type
 */
export function getTopLevelType(typePath: string): ResultTopType {
  if (!typePath) throw new Error("typePath must be a non-empty string");
  const firstSegment = typePath.split("/")[0];
  if (!VALID_TOP_TYPES.has(firstSegment)) {
    throw new Error(
      `Invalid top-level type: '${firstSegment}'. Must be one of: mutation, addition, deletion, artifact`,
    );
  }
  return firstSegment as ResultTopType;
}
